dvc/remote/index.py

```diff
@@ -0,0 +1,228 @@
+import logging
+import os
+import sqlite3
+import threading
+
+from dvc.state import _connect_sqlite
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteIndexNoop:
+    """No-op class for remotes which are not indexed (i.e. local)."""
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, typ, value, tbck):
+        pass
+
+    def checksums(self):
+        return []
+
+    def dir_checksums(self):
+        return []
+
+    def load(self):
+        pass
+
+    def dump(self):
+        pass
+
+    def invalidate(self):
+        pass
+
+    def replace(self, *args):
+        pass
+
+    def replace_all(self, *args):
+        pass
+
+    def update(self, *args):
+        pass
+
+    def update_all(self, *args):
+        pass
+
+
+class RemoteIndex:
+    """Class for indexing remote checksums in a sqlite3 database.
+
+    Args:
+        repo: repo for this remote index.
+        name: name for this index. Index db will be loaded from and saved to
+            ``.dvc/tmp/index/{name}.idx``.
+        dir_suffix: suffix used for naming directory checksums
+    """
+
+    INDEX_SUFFIX = ".idx"
+    VERSION = 1
+    INDEX_TABLE = "remote_index"
+    INDEX_TABLE_LAYOUT = "checksum TEXT PRIMARY KEY, " "dir INTEGER NOT NULL"
```

Comment:
f we are using a database, would it be better to have a single .dvc/tmp/index db, with appropriate tables/columns for handling multiple remotes, vs having multiple single-table .dvc/tmp/index/... dbs?

I'm not sure how common it is for users to have multiple remotes configured, but in the case where you have multiple large remotes, as it is right now we would be duplicating data across multiple (potentially large) index db files.

Reply:
@pmrowla If we are talking about storing unpacked *.dir cache files, then there doesn't seem to be a point to store them for each remote, since no matter the remote if it has *.dir present we trust that it also has all cache files from within it. So maybe it is a premature optimization for later general indexes for whole remotes.